### PR TITLE
Limit CPU and RAM for Each User Using cgroup

### LINKED ISSUE #4302: Limit CPU and RAM for Each User Using cgroup
### Describe the feature or change in detail

The proposed enhancement involves adding fields to the package modules, allowing administrators to customize CPU, RAM, and SWAP limits for each package in the package module. When a user is created within a specific package, the user's cgroup is updated during both user creation and package updates.

This feature has already been implemented and tested on the latest version of Ubuntu. However, it requires further advanced testing. Importantly, the core of HestiaCP remains unchanged; the modifications are confined to the package and user modules.

If you agree with these changes, I am prepared to submit a pull request to the repository.

### Would you like to sponsor this feature to have it implemented?

Yes

### PR DESCRIPTION
**Pull request by codebyoul**
The proposed enhancement involves adding fields to the package modules, allowing administrators to customize CPU, RAM, and SWAP limits for each package in the package module. When a user is created within a specific package, the user's cgroup is updated during both user creation and package updates.

This feature has already been implemented and tested on the latest version of Ubuntu. However, it requires further advanced testing. Importantly, the core of HestiaCP remains unchanged; the modifications are confined to the package and user modules.

If you agree with these changes, I am prepared to submit a pull request to the repository.

### CODE REVIEW

**Comment by codebyoul** [comment]
All issue related to this PR are fixed, this is how the test is done:

1. Install the stress package: sudo apt install -y stress.
2. Create two packages: one limited with 50%, 20 seconds, 1GB memory, and 1GB swap; and another unlimited without any constraints.
3. Create two users: one limited with the restricted package and another unlimited with the unrestricted package.
4. Open three terminals and connect with root, limited, and unlimited users.
5. In the root terminal, run: top
6. In the limited and unlimited user terminals, run: stress -m 1 --vm-bytes 4G --vm-keep.
7. Go to the limited package and change the CPU value or memory in the root terminal; you will see your changes immediately.

**Review by tlcd96 — COMMENTED** [review]
appears to be ok... 
Right now I don't have a test server to see if everything is really ok.

**Comment by jaapmarcus** [comment]
I have tested on a Debian server but can't get it working for now..

**Comment by codebyoul** [comment]
> I have tested on a Debian server but can't get it working for now..

Are you sure the cgroup conf is enabled ? can you check if the hestio.conf contain: RESOURCES_LIMIT = 'yes'

**Comment by codebyoul** [comment]
@jaapmarcus : Should the package options be displayed only when cgroup is enabled ("RESOURCES_LIMIT"="yes") or not?

File: install/hst-install-ubuntu.sh [hunk 1]
```diff
@@ -2422,10 +2433,4 @@ $HESTIA/bin/v-add-user-notification "$username" 'Welcome to Hestia Control Panel
...
 
 if [ "$interactive" = 'yes' ]; then
-	echo "[ ! ] IMPORTANT: The system will now reboot to complete the installation process."
-	read -n 1 -s -r -p "Press any key to continue"
-	reboot
-else
-	echo "[ ! ] IMPORTANT: You must restart the system before continuing!"
-fi
-# EOF
+	echo "[ ! ] IMPORTANT: The sys
```
Comment (jaapmarcus) [inline install/hst-install-ubuntu.sh]: I don't think is correct :)

File: install/hst-install-debian.sh [hunk 2]
```diff
@@ -2442,17 +2453,4 @@ cat $tmpfile
 rm -f $tmpfile
 
 # Add welcome message to notification panel
-$HESTIA/bin/v-add-user-notification "$username" 'Welcome to Hestia Control Panel!' '<p>You are now ready to begin adding <a href="/add/user/">user accounts</a> and <a href="/add/web/">domains</a>. For help and assistance, <a href="https://hestiacp.com/docs/" target="_blank">view the documentation</a> or <a href="https://forum.hestiacp.com/" target="_blank">visit our forum</a>.</p><p>Please <a href="https://github.com/hestiacp/hestiacp/issues" target="_blank">report any issues via GitHub</a>.</p><p class="u-text-bold">Have a wonderful day!</p><p><i class="fas fa-heart icon-red"></i> The Hestia Control Panel development team</p>'
-
-# Clean-up
-# Sort final configuration file
-sort_config_file
```
Comment (jaapmarcus) [inline install/hst-install-debian.sh]: We should still keep this part

File: install/hst-install-ubuntu.sh (hunk 1 above)
Comment (jaapmarcus) [reply install/hst-install-ubuntu.sh]: Same here we still want to keep the option to reboot or not

**Review by jaapmarcus — CHANGES_REQUESTED** [review]

File: package.json [hunk 3]
```diff
@@ -15,7 +15,8 @@
 		"lint-staged": "lint-staged",
 		"format": "prettier --cache --write .",
 		"preinstall": "npx only-allow npm",
-		"prepare": "husky"
+		"prepare": "husky",
+		"hestia": "cd src && ./hst_autocompile.sh --hestia --install '~localsrc'"
```
Comment (AlecRust) [inline package.json]: This feels like an unrelated change. If we add this it should be included in the development docs etc. it's probably better in a separate PR if you're keen on this.

**Comment by jaapmarcus** [comment]
> @jaapmarcus : Should the package options be displayed only when cgroup is enabled ("RESOURCES_LIMIT"="yes") or not?

Only when it is enabled yes

File: install/hst-install-debian.sh (hunk 2 above)
Comment (codebyoul) [reply install/hst-install-debian.sh]: It's my IDE mistake it's fixed

File: package.json (hunk 3 above)
Comment (codebyoul) [reply package.json]: OK i will remove it for now

**Comment by codebyoul** [comment]
All issues have been resolved. In the future, there may be additional cgroup fields that could be introduced. Consequently, it is recommended to retain the current table of packages without any new cgroup fields for better display.

**Review by codebyoul — COMMENTED** [review]
All issues are fixed

**Comment by johnnyq** [comment]
was this tested on an LXC VPS?
Im not sure if cgroups are handled well on LXC

**Comment by jaapmarcus** [comment]
LXC can't handle cgroups same as with quatas and some other things..
Can't prevent that... And it is a shame to not merge it because LXC is not supported... 

We might need to test it to make sure it doesn't break anything when it get enabled but we should not merge because LXC can't handle it ...

**Comment by jaapmarcus** [comment]
I can't get it working with Hetzner Cloud:

For the limited user:
Package / User.conf:
CPU_QUOTA='10%'
CPU_QUOTA_PERIOD='10s'
MEMORY_LIMIT='512M'
SWAP_LIMIT='512M'

hestia.conf 
RESOURCES_LIMIT='yes'

And with I run v-rebuild-user xxx 

v-update-user-cgroup gets triggered and the files are created in /etc/systemd/system.control folder 

But if I  run

`stress -m 1 --vm-bytes 3GB --vm-keep
`
Both users go to 100% CPU core usage

If you would like to have system access send me an email with your ssh key and I can give you access. It is a test server anyways.

**Comment by codebyoul** [comment]
> was this tested on an LXC VPS? Im not sure if cgroups are handled well on LXC

You have any ready LXC container with hestiacp installed ?

**Comment by jaapmarcus** [comment]
> > was this tested on an LXC VPS? Im not sure if cgroups are handled well on LXC
> 
> You have any ready LXC container with hestiacp installed ?

Can setup one easy with out any issues...

**Comment by codebyoul** [comment]
> I can't get it working with Hetzner Cloud:
> 
> For the limited user: Package / User.conf: CPU_QUOTA='10%' CPU_QUOTA_PERIOD='10s' MEMORY_LIMIT='512M' SWAP_LIMIT='512M'
> 
> hestia.conf RESOURCES_LIMIT='yes'
> 
> And with I run v-rebuild-user xxx
> 
> v-update-user-cgroup gets triggered and the files are created in /etc/systemd/system.control folder
> 
> But if I run
> 
> `stress -m 1 --vm-bytes 3GB --vm-keep ` Both users go to 100% CPU core usage
> 
> If you would like to have system access send me an email with your ssh key and I can give you access. It is a test server anyways.

I still wait for this

**Comment by codebyoul** [comment]
> was this tested on an LXC VPS? Im not sure if cgroups are handled well on LXC

I have tested in LXC container : https://linuxcontainers.org/incus/try-it/?id=72720808-e31c-4d82-a2f4-15b9a7152770#introduction, it must work if the cgroup is available, I set all properties manually and it work without issue.

**Comment by jaapmarcus** [comment]
> > I can't get it working with Hetzner Cloud:
> > For the limited user: Package / User.conf: CPU_QUOTA='10%' CPU_QUOTA_PERIOD='10s' MEMORY_LIMIT='512M' SWAP_LIMIT='512M'
> > hestia.conf RESOURCES_LIMIT='yes'
> > And with I run v-rebuild-user xxx
> > v-update-user-cgroup gets triggered and the files are created in /etc/systemd/system.control folder
> > But if I run
> > `stress -m 1 --vm-bytes 3GB --vm-keep ` Both users go to 100% CPU core usage
> > If you would like to have system access send me an email with your ssh key and I can give you access. It is a test server anyways.
> 
> I still wait for this

See reply ...

**Comment by codebyoul** [comment]
@jaapmarcus: The issue in the Hetzner Cloud doesn't seem directly related to the HestiaCP PR. It appears to be a system-level problem. I attempted to manually set the property in the system, confirming that the value is correctly configured. However, when connecting with the relevant user, the set value doesn't seem to take effect. I've opened an issue in the systemd project to request assistance with this problem [#31542](https://github.com/systemd/systemd/issues/31542)

**Comment by jaapmarcus** [comment]
You need to login as the user instead with su ..  Then it works "Fine"

**Comment by codebyoul** [comment]
If it works, then we can merge this pull request.

**Review by jaapmarcus — APPROVED** [review]
